f3/granite.go

```diff
@@ -157,36 +157,39 @@ func newRoundState(powerTable PowerTable) *roundState {
 	}
 }
 
-func (i *instance) Start() {
+func (i *instance) Start() error {
 	i.beginQuality()
-	i.drainInbox()
+	return i.drainInbox()
 }
 
 // Receives a new acceptable chain and updates its current acceptable chain.
 func (i *instance) receiveAcceptable(chain ECChain) {
 	i.acceptable = chain
 }
 
-func (i *instance) Receive(msg *GMessage) {
+func (i *instance) Receive(msg *GMessage) error {
 	if i.terminated() {
-		panic("received message after decision")
+		return fmt.Errorf("received message after decision")
 	}
 	if len(i.inbox) > 0 {
-		panic("received message while already processing inbox")
+		return fmt.Errorf("received message while already processing inbox")
 	}
 
 	// Enqueue the message for synchronous processing.
 	i.enqueueInbox(msg)
-	i.drainInbox()
+	return i.drainInbox()
 }
 
-func (i *instance) ReceiveAlarm(_ string) {
-	i.tryCompletePhase()
+func (i *instance) ReceiveAlarm(_ string) error {
+	err := i.tryCompletePhase()
+	if err != nil {
```

Comment:
Nit: use the shorthand
`if err := i.tryCompletePhase(); err != nil { ... }`

(elsewhere too)

Reply:
✓